enhancements/gep/001-loadbalancer/README.md

```diff
@@ -0,0 +1,119 @@
+        - [Story 2](#story-2)
+    - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+    - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+    - [Test Plan](#test-plan)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+    - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+- [Implementation History](#implementation-history)
+
+<!-- /toc -->
+
+## Summary
+The main goal of this proposal is to replace dns balancing with a software loadbalancer.
+Thus, all requests to the kube-apiserver from Kubernetes components (kubelet, scheduler, control-manager, kube-proxy)
+will go through this loadbalancer. Also, requests to the docker registry must also go through the loadbalancer.
+
+## Motivation
+At the moment, Gravity is using CoreDNS to determine which master to send requests to,
+but DNS is not a good tool for balancing traffic.
+This proposal aims to provide the following improvements:
+  1. Evenly distribute the load between all apiserver replicas
+  1. Improve the stability and scalability of the cluster
+  1. Nodes will react faster to master failures
+  1. Cluster upgrades will be faster and more efficient
+
+### Goals
+When creating a cluster, the user can choose the load balancer to be either external or internal.
+External loadbalancer is not controlled or configured by Gravity, it is the user's responsibility.
+The internal loadbalancer, on the contrary, is controlled and configured by Gravity, the planet-agent is responsible for this.
+When using an internal loadbalancer, a loadbalancer (HAProxy) will be installed for each node in the cluster including the master nodes.
+When adding/removing master nodes, the agent on each node should automatically reflect this in the local loadbalancer configuration.
+All Kubernetes components (kubelet, scheduler, control-manager, kube-proxy) will use the loadbalancer to access the apiserver.
+
+### Non-Goals
+
+## Proposal
+Change DNS balancer to a software loadbalancer (HAProxy).
+CoreDNS will resolve `leader.telekube.local` to either the external address (domain name or IP) of the loadbalancer (subject to which configuration?) or to `127.0.0.1` when using the internal loadbalancer.
+Docker image references should continue pointing to internal registry at `leader.telekube.local:5000`.
+To achieve this, docker registry will use port 5001 with port 5000 left to the loadbalancer.
+
+### User Stories
+#### Story 1
+As a cluster administrator, I can configure the loadbalancer when creating a cluster. If I set the type of loadbalancer to external, I can provide the external address in configuration.
+#### Story 2
+As a cluster administrator, I can change the loadbalancer settings on a running cluster.
+
+### Risks and Mitigations
+As the `leader.telekube.local` domain will not point to a master node, the applications using this address should be updated accordingly.
```

Comment:
If you're going to support changing this name, it's going to take more work as this is also part of the mTLS authentication between nodes.